src/containers/sound-editor.jsx

```diff
@@ -162,52 +236,212 @@ class SoundEditor extends React.Component {
+        const trimStart = this.state.trimStart === null ? 0.0 : this.state.trimStart;
+        const trimEnd = this.state.trimEnd === null ? 1.0 : this.state.trimEnd;
+
+        // Offline audio context needs at least 2 samples
+        if (this.audioBufferPlayer.buffer.length < 2) {
+            return;
+        }
+
+        const effects = new AudioEffects(this.audioBufferPlayer.buffer, name, trimStart, trimEnd);
+        effects.process((renderedBuffer, adjustedTrimStart, adjustedTrimEnd) => {
             const samples = renderedBuffer.getChannelData(0);
             const sampleRate = renderedBuffer.sampleRate;
             const success = this.submitNewSamples(samples, sampleRate);
-            if (success) this.handlePlay();
+            if (success) {
+                if (this.state.trimStart === null) {
+                    this.handlePlay();
+                } else {
+                    this.setState({trimStart: adjustedTrimStart, trimEnd: adjustedTrimEnd}, this.handlePlay);
+                }
+            }
         });
     }
+    getUndoItem () {
+        return {
+            ...this.copyCurrentBuffer(),
+            trimStart: this.state.trimStart,
+            trimEnd: this.state.trimEnd
+        };
+    }
     handleUndo () {
-        this.redoStack.push(this.copyCurrentBuffer());
-        const {samples, sampleRate} = this.undoStack.pop();
+        this.redoStack.push(this.getUndoItem());
+        const {samples, sampleRate, trimStart, trimEnd} = this.undoStack.pop();
         if (samples) {
             this.submitNewSamples(samples, sampleRate, true);
-            this.handlePlay();
+            this.setState({trimStart: trimStart, trimEnd: trimEnd}, this.handlePlay);
         }
     }
     handleRedo () {
-        const {samples, sampleRate} = this.redoStack.pop();
+        const {samples, sampleRate, trimStart, trimEnd} = this.redoStack.pop();
         if (samples) {
-            this.undoStack.push(this.copyCurrentBuffer());
+            this.undoStack.push(this.getUndoItem());
             this.submitNewSamples(samples, sampleRate, true);
-            this.handlePlay();
+            this.setState({trimStart: trimStart, trimEnd: trimEnd}, this.handlePlay);
+        }
+    }
+    handleCopy () {
+        this.copy();
+    }
+    copy (callback) {
+        const trimStart = this.state.trimStart === null ? 0.0 : this.state.trimStart;
+        const trimEnd = this.state.trimEnd === null ? 1.0 : this.state.trimEnd;
+
+        const trimStartSamples = trimStart * this.props.samples.length;
+        const trimEndSamples = trimEnd * this.props.samples.length;
+
+        const newCopyBuffer = this.copyCurrentBuffer();
+        newCopyBuffer.samples = newCopyBuffer.samples.slice(trimStartSamples, trimEndSamples);
+
+        this.setState({
+            copyBuffer: newCopyBuffer
+        }, callback);
+    }
+    handleCopyToNew () {
+        this.copy(this.copyToNew);
+    }
+    copyToNew () {
+        WavEncoder.encode({
```

Comment:
Can we see if we can extract the other place we do this (sound recorder) and reuse the code

Reply:
okay, gave this a try